lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -774,74 +824,62 @@ static bool canReplaceCopiedSelf(FullApplySite Apply,
-  }
-  // Bail out if any non-self arguments or indirect result that refer to the
-  // OpenedArchetype. The following optimization substitutes all occurrences of
-  // OpenedArchetype in the function signature, but will only rewrite the self
-  // operand.
-  //
-  // Note that the language does not allow Self to occur in contravariant
-  // position. However, SIL does allow this and it can happen as a result of
-  // upstream transformations. Since this is bail-out logic, it must handle all
-  // verifiable SIL.
-  for (auto Arg : Apply.getArgumentsWithoutSelf()) {
-    if (Arg->getType().getASTType().findIf([&CEI](Type t) -> bool {
-          return t->isEqual(CEI.OpenedArchetype);
-        })) {
-      return nullptr;
-    }
-  }
-  // The apply can only be rewritten in terms of the concrete value if it is
-  // legal to pass that value as the self argument.
-  if (CEI.isCopied && (!CEI.InitExistential ||
-                       !canReplaceCopiedSelf(Apply, CEI.InitExistential, DA))) {
-    return nullptr;
-  }
+  // Ensure that the callee is polymorphic.
+  assert(Apply.getOrigCalleeType()->isPolymorphic());
 
-  // Create a set of arguments.
+  // Create the new set of arguments to apply including their substitutions.
+  SubstitutionMap NewCallSubs = Apply.getSubstitutionMap();
   SmallVector<SILValue, 8> NewArgs;
-  for (auto Arg : Apply.getArgumentsWithoutSelf()) {
-    NewArgs.push_back(Arg);
+  unsigned NumApplyArgs = Apply.getNumArguments();
+  bool UpdatedArgs = false;
+  for (unsigned ArgIdx = 0; ArgIdx < NumApplyArgs; ArgIdx++) {
+    llvm::SmallDenseMap<unsigned, const ConcreteExistentialInfo *, 8>::iterator
+        ArgIt = CEIs.find(ArgIdx);
+    if (ArgIt != CEIs.end()) {
+      const ConcreteExistentialInfo *CEI = ArgIt->second;
+      // Check for Arg's concrete type propagation legality.
+      if (!canReplaceArg(Apply, *CEI, ArgIdx)) {
+        NewArgs.push_back(Apply.getArgument(ArgIdx));
+        continue;
+      }
+      UpdatedArgs = true;
+      // Ensure that we have a concrete value to propagate.
+      assert(CEI->ConcreteValue);
+      NewArgs.push_back(CEI->ConcreteValue);
+      // Form a new set of substitutions where the argument is
+      // replaced with a concrete type.
+      NewCallSubs = NewCallSubs.subst(
+          [&](SubstitutableType *type) -> Type {
+            if (type == CEI->OpenedArchetype)
+              return CEI->ConcreteType;
+            return type;
+          },
+          [&](CanType origTy, Type substTy,
+              ProtocolDecl *proto) -> Optional<ProtocolConformanceRef> {
+            if (origTy->isEqual(CEI->OpenedArchetype)) {
+              assert(substTy->isEqual(CEI->ConcreteType));
+              // Do a conformance lookup on this witness requirement using the
+              // existential's conformances. The witness requirement may be a
+              // base type of the existential's requirements.
+              return CEI->lookupExistentialConformance(proto).getValue();
```

Comment:
This callback returns an Optional already, I don't think you need to do getValue() here.